server/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -358,6 +367,138 @@ public ImmutableList<Application> getAllApplications() {
     return applicationRepository.getAllApplications();
   }
 
+  @Override
+  public CompletionStage<RelevantPrograms> relevantProgramsForApplicant(long applicantId) {
+    // Note: The Program model associated with the application is eagerly loaded.
+    CompletableFuture<ImmutableSet<Application>> applicationsFuture =
+        applicationRepository
+            .getApplicationsForApplicant(
+                applicantId, ImmutableSet.of(LifecycleStage.DRAFT, LifecycleStage.ACTIVE))
+            .toCompletableFuture();
+    ImmutableList<ProgramDefinition> activePrograms =
+        versionRepository.getActiveVersion().getPrograms().stream()
+            .map(Program::getProgramDefinition)
+            .filter(pdef -> pdef.displayMode().equals(DisplayMode.PUBLIC))
+            .collect(ImmutableList.toImmutableList());
+
+    return applicationsFuture.thenApplyAsync(
+        applications -> {
+          checkForDuplicateDrafts(applications);
+          return relevantProgramsForApplicant(activePrograms, applications);
+        },
+        httpExecutionContext.current());
+  }
+
+  private RelevantPrograms relevantProgramsForApplicant(
+      ImmutableList<ProgramDefinition> activePrograms, ImmutableSet<Application> applications) {
+    ImmutableMap<String, ProgramDefinition> activeProgramNames =
+        ImmutableMap.copyOf(
```

Comment:
I should add a comment for this. I explicitly collected to the ImmutableMap to avoid a case where the activePrograms stream contained multiple entries with the same adminName. Shouldn't happen in practice, but did in my unit tests. The ImmutableMap collector will fail since immutablemaps can't be built with duplicate keys. Here, the last one in would win.

Reply:
Sorry didn't realize you were maintaining existing functionality.

hmmmm, that's rough.  Ok let's leave it, the comment helps suss out the oddness the existing code was trying to handle.


( I think the existing code would show the duplicate name as two programs oddly enough...)